### ISSUE TITLE
App crashes when bookmark cell is deleted

### ISSUE THREAD

**Issue by hsuanhsuantsai** [uchicago-mobi/MPCS51030-2017-Winter-Forum#116]
Hi,

My app crashes when I try to delete one of my bookmark cell.
I found it comes from that my return value of function about numberOfRowsInSection does not update to the newest value. However, I have no idea how to solve it.

Does anyone know how to deal with it?

Also, I've put NetworkActivityIndicatorVisible into my code and see nothing happens. I was wondering if the web loads fast so that I can't see it on the status bar.
My UIActivityIndicator doesn't function well, either. Nothing happens with .startAnimating()
I've commit my work to the repo. Could someone take a look at my code?

Thanks.

**Comment by aaggarw99** [comment]
Are you reloading your table after a cell is removed with the updated favorites array?

**Comment by hbennett766** [comment]
It's possible that your web view is loading too fast to see the network activity indicator or your ui activity indicator. To test if they're working, set a breakpoint somewhere between where you set them to visible/startAnimating and where you hide/stop animating. It will pause the app, but if the network activity indicator is working, you'll see it spinning in the status bar, and if the ui activity indicator is working, you'll see it spinning on your screen.

Before I got to the crash you mentioned, I found a bug: if no articles are bookmarked, and you navigate to the bookmarks page, there's no way to get back out of it. I'm testing on iPhone, but check this for iPad as well. Think about how you could get a navigation bar, or a back button, to show up on this page so that users don't get stuck here or are forced to click on a bookmarked article to leave the page.

In regards to your crash, @aaggarw99 is on the right track. Check out your override method for the tableView editing style, where you support delete. First you call tableView.deleteRows(at: , with: ), _then_ you update the data source that you're using to populate your tableView, the bookmarked urls array. So by this logic, the table view now has one less row, but the array still has the old number of rows. What if you reversed these two tasks? If you first delete the url from the array, your tableView has yet to be aware anything's supposed to change, but the data source is updated with the new information, including the new number of urls. Then when you delete the row from the table, it says "ok cool, this is the same number as the array.count I'm supposed to have for the number of rows in this section."

**Comment by hsuanhsuantsai** [comment]
@aaggarw99 @hbennett766 Thank you for the help.

I just thought the function about numberOfRowsInSection should come after my delete function is finished and there should be no difference with the ordering of either deleting the cell or refreshing my array first in the delete function. That's a wrong assumption. (After setting the breakpoint, I found the function about numberOfRowsInSection came out first, that's why I felt confused.) Now it's fixed.

As for the web indicator, I had set breakpoints in the function containing the .startAnimating() method; however, the function was never executed. More interesting, the function containing .stopAnimating() method was executed. (I have the same prototpye for both the functions.)

Thanks
